pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1094 @@
+                    # happen for something like a constraint that requires that
+                    # an interval var specified by indirection has to be
+                    # present.)
+                    cpx_model.add(expr[1] == True)
+                else:
+                    cpx_model.add(expr[1])
+
+        # That's all, folks.
+        return cpx_model, visitor.pyomo_to_docplex
+
+
+@SolverFactory.register(
+    'cp_optimizer',
+    doc='Direct interface to CPLEX CP Optimizer'
+)
+class CPOptimizerSolver(object):
+    CONFIG = ConfigDict("cp_optimizer_solver")
+    CONFIG.declare('symbolic_solver_labels', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Write Pyomo Var and Constraint names to docplex model',
+    ))
+    CONFIG.declare('tee', ConfigValue(
+        default=False,
+        domain=bool,
+        description="Stream solver output to terminal."
+    ))
+    CONFIG.declare('options', ConfigValue(
+        default={},
+        description="Dictionary of solver options."
+    ))
+
+    def __init__(self, **kwds):
+        self.config = self.CONFIG()
+        self.config.set_value(kwds)
+        if docplex_available:
+            self._solve_status_map = {
+                cp.SOLVE_STATUS_UNKNOWN: TerminationCondition.unknown,
+                cp.SOLVE_STATUS_INFEASIBLE: TerminationCondition.infeasible,
+                cp.SOLVE_STATUS_FEASIBLE: TerminationCondition.feasible,
+                cp.SOLVE_STATUS_OPTIMAL: TerminationCondition.optimal,
+                cp.SOLVE_STATUS_JOB_ABORTED: None, # we need the fail status
+                cp.SOLVE_STATUS_JOB_FAILED: TerminationCondition.solverFailure
+            }
+            self._stop_cause_map = {
+                # We only need to check this if we get an 'aborted' status, so
+                # if this says it hasn't been stopped, we're just confused at
+                # this point.
+                cp.STOP_CAUSE_NOT_STOPPED: TerminationCondition.unknown,
+                cp.STOP_CAUSE_LIMIT: TerminationCondition.maxTimeLimit,
+                # User called exit, maybe in a callback.
+                cp.STOP_CAUSE_EXIT: TerminationCondition.userInterrupt,
+                # docplex says "Search aborted externally"
+                cp.STOP_CAUSE_ABORT: TerminationCondition.userInterrupt,
+                # This is in their documentation, but not here, for some reason
+                #cp.STOP_CAUSE_UNKNOWN: TerminationCondition.unkown
+            }
+
+    @property
+    def options(self):
+        return self.config.options
+
+    # Support use as a context manager under current solver API
+    def __enter__(self):
+        return self
+
+    def __exit__(self, t, v, traceback):
+        pass
+
+    def available(self, exception_flag=True):
+        return Executable('cpoptimizer').available() and docplex_available
+
+    def license_is_valid(self):
+        # TODO: not sure how to check the CPLEX license?
+        True
```

Comment:
Shouldn't this `return self.available()`?

Reply:
Good point.  It looks like the community edition is limited to search spaces of 2^1000.  I couldn't find anything in the docplex api to directly query the version (CE vs Academic/Full).  We could just try solving a toy model (and cache the result):
```python
if CPOptimizerSolver._unrestricted_license is None:
    # Note: 140*log_2(140) == 998.1 fits in CE, 
    #     141*log_2(141) == 1006.7 does not
    x = cp.integer_var_list(141, 1, 141, "X")
    m = cp.CpoModel()
    m.add(cp.all_diff(x))
    try:
        m.solve()
        CPOptimizerSolver._unrestricted_license = True
    except docplex.cp.solver.solver.CpoSolverException:
        CPOptimizerSolver._unrestricted_license = False
return CPOptimizerSolver._unrestricted_license
```